adapters/matterfull/matterfull.go

```diff
@@ -0,0 +1,216 @@
+		ImpIDs:  openrtb_ext.GetImpIDs(imps)}, nil
+}
+
+func createBidRequest(prebidBidRequest *openrtb2.BidRequest, params *openrtb_ext.ExtImpMatterfull, imps []openrtb2.Imp) *openrtb2.BidRequest {
+	prebidBidRequest.Imp = imps
+
+	if prebidBidRequest.Site != nil {
+		siteCopy := *prebidBidRequest.Site
+		prebidBidRequest.Site = &siteCopy
+		prebidBidRequest.Site.Publisher = nil
+		prebidBidRequest.Site.Domain = ""
+	}
+	if prebidBidRequest.App != nil {
+		appCopy := *prebidBidRequest.App
+		prebidBidRequest.App = &appCopy
+		prebidBidRequest.App.Publisher = nil
+	}
+	return prebidBidRequest
+}
+
+// Builds enpoint url based on adapter-specific pub settings from imp.ext
+func (adapter *adapter) buildEndpointURL(params *openrtb_ext.ExtImpMatterfull) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{PublisherID: params.PublisherID}
+	return macros.ResolveMacros(adapter.EndpointTemplate, endpointParams)
+}
+
+// MakeBids translates Matterfull bid response to prebid-server specific format
+func (adapter *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var msg = ""
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if response.StatusCode != http.StatusOK {
+		msg = fmt.Sprintf("Unexpected http status code: %d", response.StatusCode)
+		return nil, []error{&errortypes.BadServerResponse{Message: msg}}
+
+	}
+	var bidResp openrtb2.BidResponse
+	if err := jsonutil.Unmarshal(response.Body, &bidResp); err != nil {
+		msg = fmt.Sprintf("Bad server response: %d", err)
+		return nil, []error{&errortypes.BadServerResponse{Message: msg}}
+	}
+	if len(bidResp.SeatBid) != 1 {
+		var msg = fmt.Sprintf("Invalid SeatBids count: %d", len(bidResp.SeatBid))
+		return nil, []error{&errortypes.BadServerResponse{Message: msg}}
+	}
+
+	seatBid := bidResp.SeatBid[0]
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(seatBid.Bid))
+	for bid := range iterutil.SlicePointerValues(seatBid.Bid) {
+		bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+			Bid:     bid,
+			BidType: getMediaTypeForImpID(bid.ImpID, internalRequest.Imp),
+		})
+	}
+	return bidResponse, nil
+}
+
+// getMediaTypeForImp figures out which media type this bid is for
+func getMediaTypeForImpID(impID string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	for imp := range iterutil.SlicePointerValues(imps) {
+		if imp.ID == impID && imp.Video != nil {
+			return openrtb_ext.BidTypeVideo
+		}
+	}
+	return openrtb_ext.BidTypeBanner
+}
+
+// Builder builds a new instance of the Matterfull adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	urlTemplate, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
```

Comment:
When wrapping another err, use %w
```suggestion
		return nil, fmt.Errorf("unable to parse endpoint url template: %w", err)
```